### ISSUE TITLE
MemberNotNull annotations are not processed for collection expressions and collection initializers

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#79813]
**Version Used**: 2e23a0df5bf8d49f6bf856287b0eb10de08f97d9

**Steps to Reproduce**:

Compile the following code:
```csharp
using System.Diagnostics.CodeAnalysis;
using System.Collections;

object obj = "";
C c = [obj];
_ = c.FirstItem.ToString(); // CS8602

C c2 = new() { obj };
_ = c2.FirstItem.ToString(); // CS8602

C c3 = new();
c3.Add(obj);
_ = c3.FirstItem.ToString(); // fine


class C : IEnumerable
{
  public object? FirstItem { get; private set; }

  public IEnumerator GetEnumerator() => throw null!;
  [MemberNotNull(nameof(FirstItem))] public void Add(object obj) { FirstItem ??= obj; }
}
```

**Expected Behavior**:

All forms of the `.Add` call result in the same nullable state

**Actual Behavior**:
Only the explicit `.Add` invocation handles the `[MemberNotNull]` annotation and sets non-nullable state for the `FirstItem` property.

The behavior also causes IDE to suggest incorrect simplifications that would introduce nullable warnings. In the code above `IDE0028` is reported for `c3.Add();` and suggests to simplify the initialization to `C c3 = [obj];`. Doing so would introduce a `CS8602` warning for the next line

**Comment by danielllbastos** [comment]
using System.Diagnostics.CodeAnalysis;
using System.Collections;

object obj = "";
C c = [obj];
_ = c.FirstItem.ToString(); // CS8602

C c2 = new() { obj };
_ = c2.FirstItem.ToString(); // CS8602

C c3 = new();
c3.Add(obj);
_ = c3.FirstItem.ToString(); // fine


class C : IEnumerable
{
  public object? FirstItem { get; private set; }

  public IEnumerator GetEnumerator() => throw null!;
  [MemberNotNull(nameof(FirstItem))] public void Add(object obj) { FirstItem ??= obj; }
}
